smart-contract/hyperledger-fabric/v2/java/src/main/java/hu/bme/mit/ftsrg/tpcc/registry/EntityRegistry.java

```diff
@@ -0,0 +1,113 @@
+  }
+
+  @Override
+  public <Type extends SerializableEntityInterface<Type>> void create(Context ctx, Type entity) {
+    assertNotExists(entity);
+
+    String key = getKey(ctx, entity);
+    byte[] buffer = entity.toBuffer();
+    ctx.getStub().putState(key, buffer);
+  }
+
+  @Override
+  public <Type extends SerializableEntityInterface<Type>> void update(Context ctx, Type entity) {
+    // LOGGER.info("Begin update entity ");
+    assertExists(entity);
+
+    String key = getKey(ctx, entity);
+    byte[] buffer = entity.toBuffer();
+    ctx.getStub().putState(key, buffer);
+  }
+
+  @Override
+  public <Type extends SerializableEntityInterface<Type>> void delete(Context ctx, Type entity) {
+    // LOGGER.info("Begin delete entity");
+    assertExists(entity);
+
+    String key = getKey(ctx, entity);
+    ctx.getStub().delState(key);
+  }
+
+  @Override
+  public <Type extends SerializableEntityInterface<Type>> Type read(Context ctx, Type entity) {
+    // LOGGER.info("retrieve entity");
+    String key = getKey(ctx, entity);
+    byte[] data = ctx.getStub().getState(key);
+
+    if (data == null || data.length == 0) {
+      throw new Error(
+          "Entity with key \"" + key + "\" does not exist on the ledger, thus cannot parse it");
+    }
+
+    entity.fromBuffer(data);
+    return entity;
+  }
+
+  @Override
+  public <Type extends SerializableEntityInterface<Type>> List<Type> readAll(
+      Context ctx, Type entityTemplate) {
+    List<Type> entities = new ArrayList<>();
+    String compositeKey =
+        this.ctx.getStub().createCompositeKey(entityTemplate.getType(), new String[] {}).toString();
+    Iterator<KeyValue> iterator =
+        ctx.getStub().getStateByPartialCompositeKey(compositeKey).iterator();
+    while (iterator.hasNext()) {
+      KeyValue result = iterator.next();
+      // String key = result.getKey();
+      byte[] value = result.getValue();
+      EntityFactory<Type> factory = entityTemplate.getFactory();
+      Type entity = factory.create();
+      entity.fromBuffer(value);
+      entities.add(entity);
+    }
+    return entities;
+  }
+
+  // @Override
+  // public void _dispose() {
+  //   this.cache.dispose();
+  // }
+
+  private boolean keyExists(Context ctx, String key) {
+    byte[] valueOnLedger = ctx.getStub().getState(key);
+    return valueOnLedger != null && valueOnLedger.length > 0;
+  }
+
+  public <Type extends SerializableEntityInterface<Type>> boolean exists(Type obj) {
+    return keyExists(ctx, getKey(ctx, obj));
+  }
+
+  public <Type extends SerializableEntityInterface<Type>> void assertNotExists(Type obj) {
+    if (exists(obj)) {
+      throw new Error("Entity with key \"" + getKey(ctx, obj) + "\" already exists on the ledger");
+    }
+  }
+
+  public <Type extends SerializableEntityInterface<Type>> void assertExists(Type obj) {
+    if (!exists(obj)) {
+      throw new Error("Entity with key \"" + getKey(ctx, obj) + "\" does not exist on the ledger");
+    }
+  }
```

Comment:
You will need to pass `Context ctx` parameters across these, too. All methods should get a `Context` as their first parameter. The `exists` method should pass its own `ctx` parameter down to `keyExists`.